File: AWDroper/Tree/Tree.swift
//
//  Tree.swift
//  AWDroper
//
//  Created by Zihao Arthur Wang [STUDENT] on 10/24/19.
//  Copyright © 2019 Zihao Arthur Wang [STUDENT]. All rights reserved.
//

import Foundation

indirect enum Tree<T> {
    case node(data: T, subtrees: [Tree<T>], parent: Tree<T>)
    case leaf(data: T, parent: Tree<T>)
    case none
}
typealias FileTree = Tree<String>
extension Tree where T == String {
    func toString() -> String {
        switch self {
        case .node(let data, _, let parent):
            if case .none = parent {
                return data
            }
            return parent.toString()+"/"+data
        case .leaf(let data, let parent):
            if case .none = parent {
                return data
            }
            return parent.toString()+"/"+data
        default:
            return ""
        }
    }
}

File: AWDroper/AWUDProtocol/Tree&ArrayList/Tree.cpp
//
//  Tree.cpp
//  AWDroper
//
//  Created by Zihao Arthur Wang [STUDENT] on 9/26/19.
//  Copyright © 2019 Zihao Arthur Wang [STUDENT]. All rights reserved.
//

#include "Tree.hpp"
template <typename T>
Tree<T>::Node::Node(T data) {
    this->data = data;
}

//template <typename T>
//void Tree<T>::Node::appendNode(Node* node) {
//    this->children->append(node);
//    node->parent = this;
//}

//template <typename T>
//void Tree<T>::Node::remove() {
//    this->parent->children->forEachNode([=](Node* node)->void {
//        if (node->data == this->data) {
//            node->remove();
//        }
//    });
//}
//template <typename T>
//int Tree<T>::Node::count() {
//    return this->children->count();
//}


